### PR TITLE
Initial TOTP implementation

### PR DESCRIPTION
**Pull request by Hinton**
## Type of change

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Convert the `TotpService` from `clients` repository to rust.

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)

### CODE REVIEW

File: crates/bitwarden/src/vault/totp.rs [hunk 1]
```diff
@@ -22,9 +34,244 @@ pub struct TotpResponse {
...
+impl Algorithm {
+    // Derive the HMAC hash for the given algorithm
+    fn derive_hash(&self, key: &[u8], time: &[u8]) -> Result<Vec<u8>> {
+        fn compute_digest<D: Mac>(mut digest: D, time: &[u8]) -> Vec<u8> {
+            digest.update(time);
+            digest.finalize().into_bytes().to_vec()
+        }
+
+        Ok(match self {
+            Algorithm::Sha1 => compute_digest(HmacSha1::new_from_slice(key)?, time),
```
Comment (dani-garcia) [inline crates/bitwarden/src/vault/totp.rs:71]: `Hmac::new_from_slice` can't fail, as it accepts keys of any size, so we can just use `.expect()` and make this function infallible

File: crates/bitwarden/src/vault/totp.rs [hunk 2]
```diff
@@ -22,9 +34,244 @@ pub struct TotpResponse {
...
+            secret: "".to_string(),
+        }
+    }
+}
+
+impl Totp {
+    fn derive_otp(&self, time: i64) -> Result<String> {
+        let time = time / self.period as i64;
+
+        let secret = BASE32
```
Comment (dani-garcia) [inline crates/bitwarden/src/vault/totp.rs:102]: I think the secret should be base32 decoded in the from_str implementation, with that and the change suggested in `derive_hash`, we can make this function infallible

**Review by dani-garcia — COMMENTED** [review]
Left some comments, mostly style things, and some ways to make the `derive_otp` function infallible

File: crates/bitwarden/src/vault/totp.rs [hunk 3]
```diff
@@ -22,9 +34,248 @@ pub struct TotpResponse {
...
+                        .get("secret")
+                        .map(|v| v.to_string())
+                        .ok_or(Error::Internal("Missing secret in otpauth URI"))?,
+                )?,
+            }
+        } else if key.starts_with("steam://") {
+            Totp {
+                algorithm: Algorithm::Steam,
+                digits: 5,
+                secret: decode_secret(key.strip_prefix("steam://").expect("Prefix is defined"))?,
```
Comment (dani-garcia) [inline crates/bitwarden/src/vault/totp.rs:180]: We can avoid doing the same thing twice by just destructuring the strip_prefix in the else if:
```suggestion
        } else if let Some(secret) = key.strip_prefix("steam://") {
            Totp {
                algorithm: Algorithm::Steam,
                digits: 5,
                secret: decode_secret(secret)?,
```

File: crates/bitwarden/src/vault/totp.rs [hunk 4]
```diff
@@ -22,9 +34,248 @@ pub struct TotpResponse {
...
+    period: u32,
+    secret: Vec<u8>,
+}
+
+impl Default for Totp {
+    fn default() -> Self {
+        Self {
+            algorithm: Algorithm::Sha1,
+            digits: 6,
+            period: 30,
```
Comment (dani-garcia) [inline crates/bitwarden/src/vault/totp.rs:103]: I think this default impl would be better off as three constant values: DEFAULT_ALGORITHM, DEFAULT_DIGITS, DEFAULT_PERIOD. As it is, we only ever use the Default impl to extract the values and assign them one by one anyway

Comment (Hinton) [reply crates/bitwarden/src/vault/totp.rs]: True, I agree.

**Pull request merged**